pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -316,9 +333,57 @@ func vsphereLogin(cfg *VSphereConfig, ctx context.Context) (*govmomi.Client, err
 		return nil, err
 	}
 
+	// Add retry functionality
+	c.RoundTripper = vim25.Retry(c.RoundTripper, vim25.TemporaryNetworkError(int(cfg.Global.RoundTripperCount)))
+
 	return c, nil
 }
 
+// Returns a client which communicates with vCenter.
+// This client can used to perform further vCenter operations.
+func vSphereLogin(vs *VSphere, ctx context.Context) error {
+	var err error
+	clientLock.Lock()
+	defer clientLock.Unlock()
+	if vs.client == nil {
+		vs.client, err = newClient(vs.cfg, ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	m := session.NewManager(vs.client.Client)
+	// retrieve client's current session
+	u, err := m.UserSession(ctx)
+	if err != nil {
+		if soap.IsSoapFault(err) {
+			fault := soap.ToSoapFault(err).VimFault()
+			// If the PropertyCollector is not found, the saved session for this URL is not valid
+			if _, ok := fault.(types.ManagedObjectNotFound); ok {
```

Comment:
Do we care about this specific exception? Dont we want to create newClient if either `err != nil || u == nil` condition is true?
```
if err != nil || u != nil {
   return nil; // Success
}

// Log warning
vs.client.Logout(ctx)
vs.client, err = newClient(vs.cfg, ctx)